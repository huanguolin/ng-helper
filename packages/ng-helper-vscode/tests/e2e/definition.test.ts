import { expect } from 'chai';
import { describe, it, before } from 'mocha';
import * as vscode from 'vscode';

import {
    APP_PAGES_P1_HTML_PATH,
    APP_PAGES_P4_HTML_PATH,
    BAR_FOO_COMPONENT_HTML_PATH,
    DEFINITION_COMMAND,
    DRAG_SOURCE_COMPONENT_TS_PATH,
} from '../testConstants';
import { activate, setupChaiSnapshotPlugin, sleep } from '../testUtils';

describe('Definition', () => {
    setupChaiSnapshotPlugin();

    before(async () => {
        await activate();
    });

    describe('component name/attr', () => {
        it('get definition info on component name', async () => {
            await testDefinition(BAR_FOO_COMPONENT_HTML_PATH, new vscode.Position(0, 2));
        });

        it('get definition info on component attr', async () => {
            await testDefinition(BAR_FOO_COMPONENT_HTML_PATH, new vscode.Position(0, 12));
        });
    });

    describe('directive name/attr', () => {
        it('get definition info on directive name', async () => {
            await testDefinition(BAR_FOO_COMPONENT_HTML_PATH, new vscode.Position(1, 7));
        });

        it('get definition info on directive attr', async () => {
            await testDefinition(BAR_FOO_COMPONENT_HTML_PATH, new vscode.Position(1, 20));
        });
    });

    describe('type', () => {
        it('get definition info on "ctrl" (component html)', async () => {
            await testDefinition(BAR_FOO_COMPONENT_HTML_PATH, new vscode.Position(2, 14));
        });

        it('get definition info on "(ctrl.)bar" (component html)', async () => {
            await testDefinition(BAR_FOO_COMPONENT_HTML_PATH, new vscode.Position(2, 18));
        });

        it('get definition info on "ctl" (controller html)', async () => {
            await testDefinition(APP_PAGES_P1_HTML_PATH, new vscode.Position(1, 12));
        });

        it('get definition info on "(ctl.obj.)arr" (controller html)', async () => {
            await testDefinition(APP_PAGES_P1_HTML_PATH, new vscode.Position(2, 33));
        });
    });

    describe('filter', () => {
        // ts
        it('get definition info on "status"', async () => {
            await testDefinition(BAR_FOO_COMPONENT_HTML_PATH, new vscode.Position(1, 81));
        });

        // js
        it('get definition info on "f2"', async () => {
            await testDefinition(BAR_FOO_COMPONENT_HTML_PATH, new vscode.Position(1, 86));
        });

        // in not component/controller html
        it('get definition info on "status" (not component/controller html)', async () => {
            await testDefinition(APP_PAGES_P4_HTML_PATH, new vscode.Position(3, 21));
        });
    });

    describe('inline html', () => {
        it('component name', async () => {
            await testDefinition(DRAG_SOURCE_COMPONENT_TS_PATH, new vscode.Position(29, 10));
        });

        it('component attr', async () => {
            await testDefinition(DRAG_SOURCE_COMPONENT_TS_PATH, new vscode.Position(29, 24));
        });

        it('directive name', async () => {
            await testDefinition(DRAG_SOURCE_COMPONENT_TS_PATH, new vscode.Position(30, 15));
        });

        it('directive attr', async () => {
            await testDefinition(DRAG_SOURCE_COMPONENT_TS_PATH, new vscode.Position(30, 27));
        });

        it('type', async () => {
            await testDefinition(DRAG_SOURCE_COMPONENT_TS_PATH, new vscode.Position(31, 71));
        });

        it('filter', async () => {
            await testDefinition(DRAG_SOURCE_COMPONENT_TS_PATH, new vscode.Position(30, 62));
        });
    });
});

async function testDefinition(filePath: string, position: vscode.Position, waitSeconds?: number) {
    // show the document
    await vscode.window.showTextDocument(vscode.Uri.file(filePath));
    if (waitSeconds) {
        await sleep(waitSeconds * 1000);
    }

    // get definition info
    const definitionInfoList = await vscode.commands.executeCommand<vscode.Location[]>(
        DEFINITION_COMMAND,
        vscode.Uri.file(filePath),
        position,
    );

    // assert
    expect(definitionInfoList.length).to.be.greaterThan(0);
    // 要移除处理绝对路径问题，否则测试无法跨平台。
    // 方法是移除路径 ng-helper-vscode (含)之前的部分：
    definitionInfoList.forEach((definitionInfo) => {
        const relativePath = definitionInfo.uri.toString().replace(/^.*\/ng-helper-vscode\//, '');
        definitionInfo.uri = vscode.Uri.file(relativePath);
    });
    expect(definitionInfoList).toMatchSnapshot();
}
