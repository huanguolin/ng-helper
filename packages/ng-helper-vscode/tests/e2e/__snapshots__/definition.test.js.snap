// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Definition component name/attr get definition info on component attr 1`] = `
Array [
  Object {
    "range": Array [
      Object {
        "character": 8,
        "line": 17,
      },
      Object {
        "character": 29,
        "line": 17,
      },
    ],
    "uri": Object {
      "$mid": 1,
      "path": "/tests/fixtures/app/components/baz-qux/baz-qux.component.js",
      "scheme": "file",
    },
  },
]
`;

exports[`Definition component name/attr get definition info on component name 1`] = `
Array [
  Object {
    "range": Array [
      Object {
        "character": 43,
        "line": 12,
      },
      Object {
        "character": 51,
        "line": 12,
      },
    ],
    "uri": Object {
      "$mid": 1,
      "path": "/tests/fixtures/app/components/baz-qux/baz-qux.component.js",
      "scheme": "file",
    },
  },
]
`;

exports[`Definition directive name/attr get definition info on directive attr 1`] = `
Array [
  Object {
    "range": Array [
      Object {
        "character": 16,
        "line": 9,
      },
      Object {
        "character": 37,
        "line": 9,
      },
    ],
    "uri": Object {
      "$mid": 1,
      "path": "/tests/fixtures/app/directives/number-check.directive.js",
      "scheme": "file",
    },
  },
]
`;

exports[`Definition directive name/attr get definition info on directive name 1`] = `
Array [
  Object {
    "range": Array [
      Object {
        "character": 43,
        "line": 1,
      },
      Object {
        "character": 56,
        "line": 1,
      },
    ],
    "uri": Object {
      "$mid": 1,
      "path": "/tests/fixtures/app/directives/number-check.directive.js",
      "scheme": "file",
    },
  },
]
`;

exports[`Definition filter get definition info on "f2" 1`] = `
Array [
  Object {
    "range": Array [
      Object {
        "character": 12,
        "line": 2,
      },
      Object {
        "character": 16,
        "line": 2,
      },
    ],
    "uri": Object {
      "$mid": 1,
      "path": "/tests/fixtures/app/filters/f2.js",
      "scheme": "file",
    },
  },
]
`;

exports[`Definition filter get definition info on "status" (not component/controller html) 1`] = `
Array [
  Object {
    "range": Array [
      Object {
        "character": 16,
        "line": 8,
      },
      Object {
        "character": 24,
        "line": 8,
      },
    ],
    "uri": Object {
      "$mid": 1,
      "path": "/tests/fixtures/app/filters/f1.ts",
      "scheme": "file",
    },
  },
]
`;

exports[`Definition filter get definition info on "status" 1`] = `
Array [
  Object {
    "range": Array [
      Object {
        "character": 16,
        "line": 8,
      },
      Object {
        "character": 24,
        "line": 8,
      },
    ],
    "uri": Object {
      "$mid": 1,
      "path": "/tests/fixtures/app/filters/f1.ts",
      "scheme": "file",
    },
  },
]
`;

exports[`Definition inline html component attr 1`] = `
Array [
  Object {
    "range": Array [
      Object {
        "character": 8,
        "line": 17,
      },
      Object {
        "character": 29,
        "line": 17,
      },
    ],
    "uri": Object {
      "$mid": 1,
      "path": "/tests/fixtures/app/components/baz-qux/baz-qux.component.js",
      "scheme": "file",
    },
  },
]
`;

exports[`Definition inline html component name 1`] = `
Array [
  Object {
    "range": Array [
      Object {
        "character": 43,
        "line": 12,
      },
      Object {
        "character": 51,
        "line": 12,
      },
    ],
    "uri": Object {
      "$mid": 1,
      "path": "/tests/fixtures/app/components/baz-qux/baz-qux.component.js",
      "scheme": "file",
    },
  },
]
`;

exports[`Definition inline html directive attr 1`] = `
Array [
  Object {
    "range": Array [
      Object {
        "character": 16,
        "line": 8,
      },
      Object {
        "character": 26,
        "line": 8,
      },
    ],
    "uri": Object {
      "$mid": 1,
      "path": "/tests/fixtures/app/directives/number-check.directive.js",
      "scheme": "file",
    },
  },
]
`;

exports[`Definition inline html directive name 1`] = `
Array [
  Object {
    "range": Array [
      Object {
        "character": 43,
        "line": 1,
      },
      Object {
        "character": 56,
        "line": 1,
      },
    ],
    "uri": Object {
      "$mid": 1,
      "path": "/tests/fixtures/app/directives/number-check.directive.js",
      "scheme": "file",
    },
  },
]
`;

exports[`Definition inline html filter 1`] = `
Array [
  Object {
    "range": Array [
      Object {
        "character": 16,
        "line": 8,
      },
      Object {
        "character": 24,
        "line": 8,
      },
    ],
    "uri": Object {
      "$mid": 1,
      "path": "/tests/fixtures/app/filters/f1.ts",
      "scheme": "file",
    },
  },
]
`;

exports[`Definition inline html type 1`] = `
Array [
  Object {
    "range": Array [
      Object {
        "character": 8,
        "line": 19,
      },
      Object {
        "character": 31,
        "line": 19,
      },
    ],
    "uri": Object {
      "$mid": 1,
      "path": "/tests/fixtures/app/components/drag-source/drag-source.component.ts",
      "scheme": "file",
    },
  },
]
`;

exports[`Definition type get definition info on "(ctl.obj.)arr" (controller html) 1`] = `
Array [
  Object {
    "range": Array [
      Object {
        "character": 8,
        "line": 10,
      },
      Object {
        "character": 10,
        "line": 17,
      },
    ],
    "uri": Object {
      "$mid": 1,
      "path": "/tests/fixtures/app/pages/p1/p1.ts",
      "scheme": "file",
    },
  },
]
`;

exports[`Definition type get definition info on "(ctrl.)bar" (component html) 1`] = `
Array [
  Object {
    "range": Array [
      Object {
        "character": 8,
        "line": 11,
      },
      Object {
        "character": 21,
        "line": 11,
      },
    ],
    "uri": Object {
      "$mid": 1,
      "path": "/tests/fixtures/app/components/bar-foo/bar-foo.component.ts",
      "scheme": "file",
    },
  },
]
`;

exports[`Definition type get definition info on "ctl" (controller html) 1`] = `
Array [
  Object {
    "range": Array [
      Object {
        "character": 10,
        "line": 1,
      },
      Object {
        "character": 22,
        "line": 1,
      },
    ],
    "uri": Object {
      "$mid": 1,
      "path": "/tests/fixtures/app/pages/p1/p1.ts",
      "scheme": "file",
    },
  },
]
`;

exports[`Definition type get definition info on "ctrl" (component html) 1`] = `
Array [
  Object {
    "range": Array [
      Object {
        "character": 10,
        "line": 3,
      },
      Object {
        "character": 26,
        "line": 3,
      },
    ],
    "uri": Object {
      "$mid": 1,
      "path": "/tests/fixtures/app/components/bar-foo/bar-foo.component.ts",
      "scheme": "file",
    },
  },
]
`;
